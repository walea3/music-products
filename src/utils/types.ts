export type Product = {
  id: number;
  title: string;
  description: string;
  artist: string;
  isFree: boolean;
  image1: string;
  image2: string;
  accentColor: string;
  backgroundColor: string;
  textColor: string;
  feature_order: string;
  price?: string;
};
