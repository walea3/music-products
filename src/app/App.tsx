import { BrowserRouter as Router, Switch, Route } from 'react-router-dom';

import { ProductListingPage } from 'features/listing';
import { ProductDetailPage } from 'features/detail';
import styled from 'styled-components';

function App() {
  return (
    <Main>
      <Router>
        <Switch>
          <Route path="/product-:productId">
            <ProductDetailPage />
          </Route>
          <Route path="/">
            <ProductListingPage />
          </Route>
        </Switch>
      </Router>
    </Main>
  );
}

const Main = styled.main`
  max-width: 90%;
  margin: 20px auto;
`;

export default App;
