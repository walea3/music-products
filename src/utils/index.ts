export * from './product';
export * from './useAPI';
export * from './test';
