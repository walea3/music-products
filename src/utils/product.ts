export type Product = {
  id: number;
  title: string;
  description: string;
  artist: string;
  isFree: boolean;
  image1: string;
  image2: string;
  accentColor: string;
  backgroundColor: string;
  textColor: string;
  feature_order: string;
  price?: string;
};

export const BASE_PRICE = 12.99;
export const DEFAULT_BG_COLOR = '#424242';

export let naiveProductCache: { [id: number]: Product } = {};

export const hasMismatchedData = ({ id }: Product) => mismatchedDataMap[id];

export const parseProducts = (products: Product[]) => {
  naiveProductCache = products.reduce(
    (cache, product) =>
      hasMismatchedData(product) ? { ...cache, [product.id]: product } : cache,
    naiveProductCache,
  );

  return Object.keys(naiveProductCache).length
    ? products.map(fixMismatch)
    : products;
};

export const fixMismatch = (product: Product) => {
  const mismatches = mismatchedDataMap[product.id];

  return mismatches
    ? mismatches.reduce((fixedProduct, [mismatchId, ...fields]) => {
        const matchedProduct = naiveProductCache[mismatchId];
        return fields.reduce(
          (acc, field) => ({ ...acc, [field]: matchedProduct[field] }),
          fixedProduct,
        );
      }, product)
    : product;
};

const mismatchedDataMap: {
  [id: number]: [mismatchId: number, ...fields: (keyof Product)[]][];
} = {
  111: [[114, 'image1', 'image2']],
  114: [[111, 'image1', 'image2']],
};
