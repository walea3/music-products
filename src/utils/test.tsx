import { ReactElement } from 'react';
import { render } from '@testing-library/react';
import { Route, RouteProps, BrowserRouter } from 'react-router-dom';

export const renderWithRouter = (ui: ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: BrowserRouter });
};

export const renderPage = (
  page: RouteProps['children'],
  path: RouteProps['path'],
  route: string,
) => renderWithRouter(<Route path={path}>{page}</Route>, { route });

// based on react-testing library default
const TEST_ID_ATTR = 'data-testid';
const TEST_ID_DELIMITER = '-';

// @TODO if desired remove test id in prod
export const testIdProp = (baseTestId: string) => (id?: string) => ({
  [TEST_ID_ATTR]: `${baseTestId}${id ? `${TEST_ID_DELIMITER}${id}` : ''}`,
});

export const getTestId = (...args: any[]) =>
  args.filter(Boolean).join(TEST_ID_DELIMITER);
