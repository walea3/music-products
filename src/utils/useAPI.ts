import { useState, useEffect, useCallback } from 'react';

export const useAPI = <T>(url: string) => {
  const [error, setError] = useState(null);
  const [data, setData] = useState<T>();
  const [isLoading, setLoading] = useState(true);

  const fetchData = useCallback(async () => {
    setLoading(true);
    setError(null);

    try {
      const res = await fetch(`/api/${url}`);
      const data: T = await res.json();

      setData(data);
    } catch (error) {
      setError(error);
    } finally {
      setLoading(false);
    }
  }, [url]);

  useEffect(() => {
    fetchData();
  }, [fetchData]);

  return [isLoading, data, error, fetchData] as const;
};
