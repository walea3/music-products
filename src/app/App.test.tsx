import React from 'react';

import { renderWithRouter } from 'utils';

import App from './App';

describe('<App />', () => {
  describe('@renders', () => {
    test('initial route as product listings page', () => {
      const wrapper = renderWithRouter(<App />);
      expect(wrapper.asFragment()).toMatchSnapshot();
    });

    test('product route with product id', () => {
      const wrapper = renderWithRouter(<App />, { route: '/product-1' });
      expect(wrapper.asFragment()).toMatchSnapshot();
    });
  });
});
