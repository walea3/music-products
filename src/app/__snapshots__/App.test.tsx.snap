// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> @renders initial route as product listings page 1`] = `
<DocumentFragment>
  .c0 {
  max-width: 90%;
  margin: 20px auto;
}

<main
    class="c0"
  >
    <div>
      <h1>
        Products
      </h1>
      <div>
        Loading products ...
      </div>
      <div>
        No products, yet.
      </div>
    </div>
  </main>
</DocumentFragment>
`;

exports[`<App /> @renders product route with product id 1`] = `
<DocumentFragment>
  .c0 {
  max-width: 90%;
  margin: 20px auto;
}

<main
    class="c0"
  >
    <div>
      Loading product ...
    </div>
  </main>
</DocumentFragment>
`;
